File: CivitasRuby/MazoSorpresas.rb
module Civitas

class MazoSorpresas

  def init
    @sorpresas =[]
    @cartasEspeciales=[]
    @barajadas=false
    @usadas=0

  end

  def initialize(debug=false)
    @debug=debug
    init
    if debug 
      Diario.ocurreEvento('El mazo de sorpresas se encuentra en modo debug.')
    end
  end

  def alMazo(s)
    if !@barajadas
      @sorpresas.push(s)
    end
  end

  def siguiente
    if (!@barajadas and @usadas==@sorpresas.length and !@debug)
      @sorpresas.shuffle
      @usadas=0
      @barajadas=true
    end
    @usadas+=1
    @ultimaSorpresa=@sorpresas[0]
    @sorpresas.push(@ultimaSorpresa)
    @sorpresas.shift
    return @ultimaSorpresa

  end

  def inhabilitarCartaEspecial(sorpresa)
    if @sorpresas.include? sorpresa
      @cartasEspeciales.push(sorpresa)
      @sorpresas.delete(sorpresa)
      Diario.instance.ocurre_evento('Se ha inhabilitado una sorpresa.')
    end

  end

  def habilitarCartaEspecial(sorpresa)
    if @cartasEspeciales.include? sorpresa
      @sorpresas.push(sorpresa)
      @cartasEspeciales.delete(sorpresa)
      Diario.instance.ocurre_evento('Se ha habilitado una sorpresa.')
    end

  end

  private :init

end

end

File: CivitasJava/civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {

    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;

    private void init() {
        this.barajada = false;
        this.usadas = 0;
        this.sorpresas = new ArrayList<>();
        this.cartasEspeciales = new ArrayList<>();
    }

    MazoSorpresas(boolean debug) {
        this.debug = debug;
        init();

        if (debug) {
            Diario.getInstance().ocurreEvento("El mazo de sorpresas se encuentra en modo debug.\n");
        }
    }

    MazoSorpresas() {
        this.debug = false;
        init();
    }

    void alMazo(Sorpresa s) {
        if (!this.barajada) {
            sorpresas.add(s);
        }
    }

    Sorpresa siguiente() {
        if ((!this.barajada || this.usadas == this.sorpresas.size()) && !this.debug) {
            Collections.shuffle(this.sorpresas);
            this.barajada = true;
            this.usadas = 0;
        }

        Sorpresa ret=this.sorpresas.get(0);
        Collections.swap(this.sorpresas,0,sorpresas.size()-1);
        this.ultimaSorpresa=ret;
        this.usadas++;
        return ret;
    }

    void inhabilitarCartaEspecial(Sorpresa sorpresa) {
        if (this.sorpresas.contains(sorpresa)) {
            this.sorpresas.remove(sorpresa);
            this.cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha inhabilitado una sorpresa.\n");
        }
    }

    void habilitarCartaEspecial(Sorpresa sorpresa) {
        if (this.cartasEspeciales.contains(sorpresa)) {
            this.sorpresas.add(sorpresa);
            this.cartasEspeciales.remove(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha habilitado una sorpresa.\n");
        }
    }
}
